spec/models/queries/projects/project_query_custom_field_order_spec.rb

```diff
@@ -0,0 +1,324 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.describe ProjectQuery, "order using CustomFieldOrder" do
+  shared_let(:user) { create(:admin) }
+
+  let(:query_results) do
+    described_class.new.order(custom_field.column_name => direction, id: :asc).results.to_a
+  end
+
+  before do
+    login_as(user)
+  end
+
+  def cf_values(*values)
+    values.map do |value|
+      CustomValue.new(custom_field_id: custom_field.id, value:)
+    end
+  end
+
+  def project_with_cf_value(*values)
+    create(:public_project, custom_values: cf_values(*values))
+  end
+
+  shared_examples "it sorts" do
+    let(:projects_desc) { projects.reverse }
+
+    before { projects }
+
+    context "in ascending order" do
+      let(:direction) { :asc }
+
+      it "returns the correctly sorted result" do
+        expect(query_results.map(&:id))
+          .to eq projects.map(&:id)
+      end
+    end
+
+    context "in descending order" do
+      let(:direction) { :desc }
+
+      it "returns the correctly sorted result" do
+        expect(query_results.map(&:id))
+          .to eq projects_desc.map(&:id)
+      end
+    end
+  end
+
+  context "for string format" do
+    include_examples "it sorts" do
+      let(:custom_field) { create(:project_custom_field, :string) }
+
+      let(:projects) do
+        [
+          create(:public_project),
```

Comment:
This would read even better, if the syntactic sugar `project_without_cf_value` would be used.